### PR TITLE
OpenShift CoreOS Layering (provisional)

### PR DESCRIPTION
**Pull request by cgwalters**
**NOTE: Nothing in this proposal should be viewed as final.
It is highly likely that details will change.  It is quite possible
that larger architectural changes will be made as well.**

Change RHEL CoreOS as shipped in OpenShift to be a "base image" that can
be used as in layered container builds and then booted.  This will allow
custom 3rd party agents delivered via RPMs installed in a container
build.  The MCO will roll out and monitor these custom builds the same
way it does for the "pristine" CoreOS image today.

This is the OpenShift integration of [ostree native containers](https://fedoraproject.org/wiki/Changes/OstreeNativeContainer) or [CoreOS layering](https://github.com/coreos/enhancements/pull/7) via the MCO.

### CODE REVIEW

File: enhancements/ocp-coreos-layering.md
Comment (danwinship) [inline enhancements/ocp-coreos-layering.md:92]: @cgwalters yes, this user story exactly covers the problem I was asking about on Slack (subject to the warning down below under "Support Procedures" about making sure it's very clear when a customer is doing this)

**Comment by cgwalters** [comment]
I'd like to proceed with at least adding a new `rhcos` entry to the release payload for 4.11 (it wouldn't actually be *used* at first), and we plan to try to land some preparatory stuff for this to the MCO git main/master.  So if we can get some initial even tentative approval for the broad outlines of this proposal, that'd be great!

File: enhancements/ocp-coreos-layering.md
Comment (mrunalp) [inline enhancements/ocp-coreos-layering.md:100]: We can add a note that this could be automated in the future (detection and removal of hotfix with released rpm).

File: enhancements/ocp-coreos-layering.md
Comment (mrunalp) [inline enhancements/ocp-coreos-layering.md:131]: Could we add a section on upgrade flow? It could cover:
1. How we move existing clusters to the new format.
2. How the imagestream is updated as new OpenShift updates go out.
(The information is there in the doc, but it may be easier to follow if we have a dedicated section.)

File: enhancements/ocp-coreos-layering.md
Comment (cybertron) [inline enhancements/ocp-coreos-layering.md:254]: Okay, I think that answers my questions above. The boot image customization would be a future feature (that we would be very interested in).

**Review by cybertron — COMMENTED** [review]
Still working my way through this, but a few initial comments inline from a host networking perspective.

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:61]: Everything that works today will continue to work.  We've invested a *ton* of time on the CoreOS side into handling all kinds of network configuration, and that all will continue to work.  Specifically, static IP addresses can (and probably should) still be configured via providing Ignition on a per-machine/node basis.

File: enhancements/ocp-coreos-layering.md
Comment (c3d) [inline enhancements/ocp-coreos-layering.md:96]: Could you contrast the "before/after" situation and outline benefits / drawbacks, notably in this scenario? I'm notably concerned with the override continuing to apply. IIUC, that's a "whole kernel" override, not just a given patch, so if there is a second hotfix with a wider distribution, how would the two combine?

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:139]: OK, added a bit more about 1.

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:96]: > I'm notably concerned with the override continuing to apply.

I pushed a bit about perhaps using a Bugzilla entry to know whether or not we still need the override.

**Comment by cgwalters** [comment]
@mrunalp @aravindhp  Given this is all provisional, any further concerns before merging this?

File: enhancements/ocp-coreos-layering.md
Comment (janosdebugs) [inline enhancements/ocp-coreos-layering.md:37]: Currently, AFAIK the licensing of CoreOS does not allow reusing it for other purposes (FIXME). Will this change with this proposal?

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:37]: There is (to the best of my knowledge) no *licensing* restrictions - "RHEL CoreOS" is just not a separate thing from OpenShift.

I think this question is really addressed by https://github.com/openshift/enhancements/pull/1032/files#diff-d56db1c5d442467039e4ac4e6d118edd49185320a947906e8fd6973cbbd5f1abR60

But, the underlying non-OCP functionality of "boot a container" is in https://fedoraproject.org/wiki/Changes/OstreeNativeContainer and https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md and will ship in the `rpm-ostree` in RHEL.  Truly productizing that is a whole separate thing.  I pushed https://github.com/openshift/enhancements/pull/1032/commits/d25534f6ecb1a27d76b7f542ac871fe31599cf75 related to that.

File: enhancements/ocp-coreos-layering.md
Comment (Prashanth684) [inline enhancements/ocp-coreos-layering.md:84]: is it also possible to do it on an initial cluster install? by just doing an `openshift-install create manifests` and then adding this custom manifest with this config which MCO would rollout during install?

File: enhancements/ocp-coreos-layering.md
Comment (Prashanth684) [inline enhancements/ocp-coreos-layering.md:86]: i'm wondering how this will be supported in the context of heterogeneous clusters as the internal image registry will not support manifestlisted images and builds doesn't support it either: https://github.com/openshift/enhancements/pull/1014/files#r797900391

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:86]: Pushed a blurb on this

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:84]: We could clearly use the bootstrap node to perform a build (e.g. `podman build`).  I think performance needs will drive us to do this because otherwise we'd be adding a whole other phase to rollout where we'd need to wait for the cluster to be up to do a build, to generate our container image to roll it out to all the nodes.

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:84]: I suspect a number of administrators though will end up wanting to provide an *externally* built image; that's alluded to above.  A whole host of tradeoffs with thaat.

Comment (Prashanth684) [reply enhancements/ocp-coreos-layering.md:84]: hmm..in that case where would the builds be pushed though ? so that it can be picked up my the individual nodes' MCO?

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:84]: We'd run a registry on the bootstrap node.

Comment (Prashanth684) [reply enhancements/ocp-coreos-layering.md:84]: rigth..externally built images is definitely an option but most of the time we have requests like - "i want to add this systemd unti/file to ignition " or just change some kargs for which this will be useful

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:84]: Yep definitely, all that works via MachineConfig today and we intend to keep it working.

**Comment by mrunalp** [comment]
/approve

(will keep it open for a couple more days to see if there more comments from others before we merge)

**Review by mrunalp — APPROVED** [review]

File: enhancements/ocp-coreos-layering.md
Comment (sferich888) [inline enhancements/ocp-coreos-layering.md:68]: @control-d I think this means we can now ship this image on registry.redhat.io (or the container catalog) and not just on Quay.

Comment (sferich888) [reply enhancements/ocp-coreos-layering.md:68]: I wonder if this does or would solve some of the CVE reporting issues we have with RHCOS today?

Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:68]: Yes, a big sub-goal here is that the container image (though it has some "ostree stuff" in there) otherwise looks like e.g. the UBI base image, and should be understandable to security scanners like Clair directly.

There's some details though around how we handle extensions - those would not be visible to scanners.

File: enhancements/ocp-coreos-layering.md
Comment (bparees) [inline enhancements/ocp-coreos-layering.md:132]: so if i'm understanding it correctly, this is the spot at which a user/admin would define the image customizations they want to apply, by defining a buildconfig that points to a dockerfile that contains their additions, more or less?

picking up @fabiand's question from the mailing list:

> How is it imagined to include different packages from different vendors in the same base image?
Let's say we've got a HW vendor providing (proprietary) drivers for their GPU, hardware monitoring tools by the system vendor, and some IDS. Some of this will be provided by an Operator, other elements by a human. How are they expected to create a single new base image?

and @cgwalters answer:
> Hmm; in theory perhaps we could create an interface (like a CRD) that
allowed registering "build fragments" (much like MachineConfig) that
could all be gathered and merged into a single build.

is there another option here that we allow multiple user-provided buildconfigs, with some sort of ordering, where only the final build produces the actual coreos image that will be consumed by the cluster?  I don't know if it's strictly necessary, vs just aggregating all the customization logic into a single buildconfig/dockerfile, but i wanted to continue the thread.

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:69]: We actually today do ship some content in CoreOS that doesn't come from an RPM - systemd units and the like: https://github.com/coreos/fedora-coreos-config/tree/testing-devel/overlay.d

Doing that for binaries...we want to enable that for customers.  For us (e.g. to stop building kubelet via RPM) - definitely *possible* but has other ramifications.

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:377]: > (reimplementing a lot of push/pull secrets, proxy configs, CAs, ICSP logic)

We already need to be able to pull container images on the bootstrap node with plain `podman`, and we already do.  So that covers CAs and ICSP too.

The push aspect is interesting, but I think the most likely thing here is to run a basic in-memory registry (possibly the in-cluster one, just in a non-production mode) and we can easily set things up to allow pushing to that.  Or we just use `oc image serve` etc.

File: enhancements/ocp-coreos-layering.md
Comment (cgwalters) [reply enhancements/ocp-coreos-layering.md:132]: I feel confident that: no matter what we pick, it will have feature gaps :smile:

This enhancement specifically didn't yet try to call out the details of how we propose the user interface to work, because I think we simply need time to iterate on it, provide a functioning preview and and learn what works before we can draw a roadmap into that area.

Comment (bparees) [reply enhancements/ocp-coreos-layering.md:132]: totally fair, just trying to get some thoughts in place for when you get to that point

**Comment by mrunalp** [comment]
/lgtm

I think we have enough to merge this provisional enhancement. We can keep iterating and working through what the UI will look like and update the KEP.

**Comment by zhouhao3** [comment]
@cgwalters Hi, I want to test this feature in OCP. I noticed that one of the main tasks of this proposal is to build a container image of RHEL CoreOS. Is this image built? I didn't find this image. Thanks!

**Comment by cgwalters** [comment]
It is published but only for OCP developers right now at `registry.ci.openshift.org/rhcos-devel/rhel-coreos:latest`.

Publishing this image officially is tracked by https://url.corp.redhat.com/d4e7ed8

**Pull request merged**
